File: joonas/042.cpp
class Solution {
public:
    int h, w;
    vector<vector<char>> s;
    char v[6][6][15][5];
    bool vis[6][6];
    string word;

    char solve(int y, int x, int wi, int pd) {
        if (y < 0 || y >= h || x < 0 || x >= w || word[wi] != s[y][x] || vis[y][x]) return 0;
        if (wi + 1 >= word.length()) return 1;
        vis[y][x] = true;
        char& ret = v[y][x][wi][pd];
        if (ret == 1) {
            vis[y][x] = false;
            return ret;
        }
        ret = 0;
        const int dy[] = { 1,0,-1,0 };
        const int dx[] = { 0,-1,0,1 };
        for (int i = 0; i < 4; ++i) {
            int ny = y + dy[i], nx = x + dx[i];
            if (solve(ny, nx, wi + 1, i) == 1) {
                ret = 1;
                break;
            }
        }
        vis[y][x] = false;
        return ret;
    }

    bool exist(vector<vector<char>>& board, string _w) {
        memset(v, -1, sizeof(v));
        s = board;
        h = board.size();
        w = board[0].size();
        word = _w;
        for (int i = 0; i < h; ++i) {
            for (int j = 0; j < w; ++j) {
                if (solve(i, j, 0, 4) == 1) return true;
            }
        }
        return false;
    }
};

File: jlogkim/Season2/042.java
class Solution {
    boolean visit[][];
    char[][] sol_board;
    int way[][] = {{1,0},{-1,0},{0,1},{0,-1}};
    String target;
    private boolean hasWord(int i,int j,int bef){
        int nr,nc;
        if((bef+1)>=target.length()) return true;
        for(int idx=0; idx<4; idx++){
            nr=i+way[idx][0];
            nc=j+way[idx][1];
            if(nr>=0&&nr<sol_board.length&&nc>=0&&nc<sol_board[0].length){
                if(sol_board[nr][nc]==target.charAt(bef+1)&&visit[nr][nc]==false){
                    visit[nr][nc]=true;
                    if(hasWord(nr,nc,bef+1)) return true;
                    visit[nr][nc]=false;
                }
            }
        }
        return false;
    }
    
    public boolean exist(char[][] board, String word) { 
        visit= new boolean[board.length][board[0].length];
        sol_board = new char[board.length][board[0].length];
        target=word;
        for(int i=0;i<board.length;i++){
            for(int j=0;j<board[i].length;j++){
                sol_board[i][j]=board[i][j];       
            }
        }
        
        for(int i=0;i<board.length;i++){
            for(int j=0;j<board[i].length;j++){
                if(board[i][j]==word.charAt(0)){
                    visit[i][j]=true;
                    if(hasWord(i,j,0)) return true;
                    visit[i][j]=false;
                }        
            }
        }
        return false;
    }
}